### PR TITLE
feat: Horizontal Media List Feature - Mobile

### PR DESCRIPTION
**Pull request by calebpanza**
# Please note that this PR is specifically for the Mobile App. TV will be a separate PR

## ✏️ Description

This PR introduces the `HorizontalMediaListFeature` into the `components` directory. Please note, this PR **does not** integrate the feature into the `FeatureFeed`. This PR does the following:

- Adds the `HorizontalMediaListFeature` to the `components` directory
- Renames the `MediaCard` to `MediaItem` to better reflect the design which is _not_ a card
- Adds a `useLivestreamIsActive` hook that accepts a `Livestream` object and locally manages reporting when the stream is live or not
- Adds a `LIVE` label to the `MediaItem`

## 📸 Screenshots and 🔬 Testing
**HorizontalMediaListFeature**
[Loom](https://www.loom.com/share/18be04b8814444a6988b4233ab947d4c)

To test this, set your `PROJECT_VARIANT` to `storybook` in your `env` and start the app. You can find a new story under `Components/HorizontalMediaListFeature` with a few demos.

**MediaItem**
[Loom](https://www.loom.com/share/3ab3b7f59b454e409422ba89185f3777)

Testing this is really just gonna be focused around making sure that imports are correct. Since the component has been renamed, that's the largest risk at play. To see the new `isLive` prop in action, the stories have also been updated accordingly

**useLivestreamIsActive**
[Loom](https://www.loom.com/share/98cbdce0710643c597a8c9866404b109)

To test this, read through the code and confirm that all potential edge-cases are taken into account. Tests are split up into a couple of different "groups", including a full simulation of a live event starting and ending. Read through the test file to make sure that the tests come to logical conclusions.

## ⚠️ To-do Before Merge
n/a

## 📝 Checklist

- [x] I have performed a self-review of my code
- [x] I have tested my code works in the project (nothing breaks)
- [x] I have commented my code in hard-to-understand areas
- [x] New and existing tests and linting pass locally with my changes
- [x] I have made corresponding changes to the documentation (new envs?)
- [x] For PR's with design updates/changes, I have consulted with the product designer and made all necessary adjustments in Figma

### CODE REVIEW

File: templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+const HorizontalMediaListFeature = ({ title, subtitle, items }) => {
+  return (
+    <Styled>
+      {isNotEmpty(subtitle) ? <Subtitle>{subtitle}</Subtitle> : null}
+      {isNotEmpty(title) ? <Title>{title}</Title> : null}
+      <ItemList
+        hasTitle={isNotEmpty(title) || isNotEmpty(subtitle)}
+        data={items}
+        renderItem={renderItem}
```
Comment (redreceipt) [inline templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js:54]:
```suggestion
        renderItem={MediaItemWithLiveStreamManager}
```

File: templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  return !isEmpty(inp);
+}
+
+function renderItem({ item }) {
+  return <MediaItemWithLiveStreamManager item={item} />;
+}
+
+const MediaItemWithLiveStreamManager = ({ item }) => {
+  const isLive = useLivestreamIsActive(
+    item?.relatedNode?.__typename === 'Livestream' ? item?.relatedNode : null
```
Comment (redreceipt) [inline templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js:31]: could you just pass in the node and let the hook handle this logic? may need to extend it to other node types in the future

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      withTiming(0.65, { duration: ANIMATION_DURATION }),
+      -1,
+      true
+    );
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  return (
+    <LiveChipContainer entering={FadeIn} exiting={FadeOut}>
+      <DarkThemeProvider>
```
Comment (redreceipt) [inline templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:47]: why do we have re-include this provider in the tree here?

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  useSharedValue,
+} from 'react-native-reanimated';
+import { DarkThemeProvider } from 'ui-kit/theme';
+import { unit } from 'ui-kit/utils';
+
+import Icon from '../Icon';
+import { SystemText } from '../typography';
+import { LiveChipContainer } from './MediaItem.styles';
+
+const paddingRight = unit(0.5);
```
Comment (redreceipt) [inline templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:23]: why is this here outside of the component?

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    };
+  });
+
+  useEffect(() => {
+    opacity.value = withRepeat(
+      withTiming(0.65, { duration: ANIMATION_DURATION }),
+      -1,
+      true
+    );
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (redreceipt) [inline templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:42]: why should this effect not fire if `opacity` changes (assuming it's some kind of memoized object)?

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  }, []);
+
+  return (
+    <LiveChipContainer entering={FadeIn} exiting={FadeOut}>
+      <DarkThemeProvider>
+        <Animated.View style={iconAnimatedStyles}>
```
Comment (redreceipt) [inline templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:48]: does styled components not work with react-native-reanimated?

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        <SystemText small bold color="text.primary">
+          LIVE
+        </SystemText>
+      </DarkThemeProvider>
+    </LiveChipContainer>
+  );
+};
+
+LiveChip.propTypes = {};
+LiveChip.defaultProps = {};
```
Comment (redreceipt) [inline templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:60]:
```suggestion
```

File: templates/mobile/src/ui-kit/components/MediaItem/MediaItem.js [hunk 8]
```diff
@@ -12,22 +12,25 @@ import isEmpty from 'lodash/isEmpty';
...
 
-const MediaCard = ({
+const MediaItem = ({
   title,
   coverImage,
-  isInHorizontalList,
+  isInHorizontalList = false,
   label,
   progressBarProgress,
+  isLive = false, // todo : is there a better way to handle this that could apply styling for more chips? or should we leave `live` a separate entity?
```
Comment (redreceipt) [inline templates/mobile/src/ui-kit/components/MediaItem/MediaItem.js:33]: wanna set these as default props instead?

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js (hunk 3 above)
Comment (calebpanza) [reply templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:47]: In order to use the dark-mode text color, I have to nest another theme provider. This is similar to what we use to do with our ThemeMixin, but in a more "styled-component-y" way

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js (hunk 4 above)
Comment (calebpanza) [reply templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:23]: So it doesn't get calculated on each render

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js (hunk 5 above)
Comment (calebpanza) [reply templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:42]: Yeah it's causing an infinite loop if done that way. I originally tried that out cause of the linter warning.

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js (hunk 6 above)
Comment (calebpanza) [reply templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:48]: They do, but the stylesheet created by re-animated needs to be generated using a hook they provide. In cases like these (where theme references aren't needed), I've found it easier to just negate styled-components and use straight up Reanimated.

File: templates/mobile/src/ui-kit/components/MediaItem/MediaItem.js (hunk 8 above)
Comment (calebpanza) [reply templates/mobile/src/ui-kit/components/MediaItem/MediaItem.js:33]: Per @conrad-vanl this is the more "React-y" way of handling defaults, now? I've always done it using `defaultProps` but updated this based on his comments.

File: templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js (hunk 2 above)
Comment (calebpanza) [reply templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js:31]: Not quite following. The only node type that will ever be "live" is the `Livestream` type, and we're passing in the entire `Livestream` node, here

File: templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js (hunk 1 above)
Comment (calebpanza) [reply templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js]: That's cool! I didn't realize you could pass the functional component reference in there. Updated!

File: templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js (hunk 2 above)
Comment (redreceipt) [reply templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js:27]: right, so instead of `HorizontalMediaListFeature` having to keep up with which node types are valid, the hook does. We only have to know about the shape of the relatedNode and act on it in one place, not two.

File: templates/mobile/src/ui-kit/components/MediaItem/MediaItem.js (hunk 8 above)
Comment (redreceipt) [reply templates/mobile/src/ui-kit/components/MediaItem/MediaItem.js:33]: I don't buy it. Why would there be a `defaultProps` field on a React component if it shouldn't be used? It's not deprecated.

Comment (redreceipt) [reply templates/mobile/src/ui-kit/components/MediaItem/MediaItem.js:33]: @conrad-vanl explain yourself

Comment (calebpanza) [reply templates/mobile/src/ui-kit/components/MediaItem/MediaItem.js:33]: I'll revert it if it gets this PR merged. Lol. I don't have a strong opinion about it

File: templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js (hunk 2 above)
Comment (calebpanza) [reply templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js:27]: OH! I see what you're saying, now. Ok yes. One moment

Comment (calebpanza) [reply templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js:27]: So sorry! That one took me a second

Comment (calebpanza) [reply templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js:27]: No other node types use the `start` and `durationInSeconds` so I think I can literally just kill that whole check

**Comment by calebpanza** [comment]
@redreceipt lmk if this is good to go, now, and I'll merge and start working on TV

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js (hunk 4 above)
Comment (redreceipt) [reply templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:23]: I'm confused. We use this unit function everywhere but we don't use global variables everywhere. Styles should only ever be computed once. That's the point of style sheets, do you don't declare global variables. Reanimated even recommends using a style sheets. We should too.

Screenshot 2023-02-07 at 4 49 19 PM

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js (hunk 5 above)
Comment (redreceipt) [reply templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:42]: that tells me that opacity is changing often and we probably shouldn't ignore or else we'll have unexpected results here. that linter warning is there for a good reason.

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js (hunk 3 above)
Comment (redreceipt) [reply templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:47]: We nest theme mixin when we want the theme to change, not to simply redeclare the theme. Why should we list the same context provider over and over. that defeats the purpose of react context.

File: templates/mobile/src/ui-kit/components/MediaItem/MediaItem.js (hunk 8 above)
Comment (redreceipt) [reply templates/mobile/src/ui-kit/components/MediaItem/MediaItem.js:33]: I do. I think we should use default props here.

File: templates/mobile/src/utils/useLivestreamIsActive.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  useEffect(() => {
+    if (!start || !durationInSeconds) {
+      return () => null;
+    }
+
+    /**
+     * Parsing ISO strings can be tricky and date-fns throws errors rather un-gracefully
+     * Instead of writing a bunch of conditions to take all parsing inconsistencies into account, we'll just use a try/catch to make sure we don't crash the whole app over the LIVE banner
+     */
+    try {
```
Comment (redreceipt) [inline templates/mobile/src/utils/useLivestreamIsActive.js:33]: where are we expecting to get an error here? I don't like wrapping whole blocks of code with try/catch blocks because we miss unrelated errors down the line. We should identify where we think the error will come from, try to mitigate, and if we're still worried, only then wrap that single line in a try/catch

Comment (redreceipt) [reply templates/mobile/src/utils/useLivestreamIsActive.js:33]: also in dev, we'd never know what was wrong or be alerted

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js (hunk 3 above)
Comment (calebpanza) [reply templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:47]: This is the way that Styled Components recommends doing it. It's how we're handling light/dark switching and even with the Pressable(Button/Icon/Chip), each variant is its own "provider". I see what you're saying and wonder if it's a nuance in the language they chose to use, but their documentation recommends this approach: https://styled-components.com/docs/advanced#function-themes

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js (hunk 5 above)
Comment (calebpanza) [reply templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:42]: You're right. I just added it back to the dependency list and it's working just fine, though. I guess early on there must have been something triggering a re-render or something but it's working just fine, now, without the linter exception

File: templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js (hunk 1 above)
Comment (calebpanza) [reply templates/mobile/src/components/HorizontalMediaListFeature/HorizontalMediaListFeature.js]: Reverted this cause you can't render hooks outside of a functional _component_ and that renderItem method runs a function instead of rendering a component.

File: templates/mobile/src/utils/useLivestreamIsActive.js (hunk 9 above)
Comment (calebpanza) [reply templates/mobile/src/utils/useLivestreamIsActive.js]: fixed!

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js (hunk 3 above)
Comment (redreceipt) [reply templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:42]: I just realized what `DarkThemeProvider` is. It's not really a provider in the sense that it's providing context. It's just overriding the theme. We're just forcing dark mode here. Shouldn't it be driven off the users device settings?

**Review by redreceipt — APPROVED** [review]

File: templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js (hunk 3 above)
Comment (calebpanza) [reply templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:42]: This is something that I need to refine in the future, but not right now. From the system theme, the text color will switch back and forth, but I only want the nearly-white text color which lives exclusively in the dark theme.

Comment (conrad-vanl) [reply templates/mobile/src/ui-kit/components/MediaItem/LiveChip.js:42]: said a different way is the chip itself is always "Dark Mode" regardless of device setting
